be/src/exprs/quantile_function.cpp

```diff
@@ -0,0 +1,127 @@
+            quantile_state.set_compression(compression);
+        }
+    }
+    
+    if(!src.is_null) {
+        StringParser::ParseResult parse_result = StringParser::PARSE_SUCCESS;
+        double double_value = StringParser::string_to_float<double>(
+                reinterpret_cast<char*>(src.ptr), src.len, &parse_result);
+        if (UNLIKELY(parse_result != StringParser::PARSE_SUCCESS)) {
+            std::stringstream error_msg;
+            error_msg << "The input: " << std::string(reinterpret_cast<char*>(src.ptr), src.len)
+                      << " is not valid, to_bitmap only support bigint value from 0 to "
+                         "18446744073709551615 currently";
+            ctx->set_error(error_msg.str().c_str());
+            return StringVal::null();
+        }
+        quantile_state.add_value(double_value);
+    }
+    return serialize(ctx, &quantile_state);
+}
+
+void QuantileStateFunctions::quantile_union(FunctionContext* ctx, const StringVal& src, StringVal* dst) {
+    if(src.is_null) {
+        return;
+    }
+    auto dst_quantile = reinterpret_cast<QuantileState<double>*>(dst->ptr);
+    if(src.len == 0) {
+        dst_quantile->merge(*reinterpret_cast<QuantileState<double>*>(src.ptr));
+    } else {
+        QuantileState<double> state(Slice(src.ptr, src.len));
+        dst_quantile->merge(state);
+    }
+}
+
+DoubleVal QuantileStateFunctions::quantile_percent(FunctionContext* ctx, StringVal& src) {
+    const AnyVal* percentile = ctx->get_constant_arg(1);
+    if (percentile != nullptr) {
+        float percentile_value = reinterpret_cast<const FloatVal*>(percentile)->val;
+        if (percentile_value > 1 || percentile_value <0)
+        {
+            std::stringstream error_msg;
+            error_msg<< "The percentile must between 0 and 1, but input is:" << std::to_string(percentile_value);
+            ctx->set_error(error_msg.str().c_str());
+            return DoubleVal::null();
+        }
+        
+        if (src.len == 0) {
+            auto quantile_state = reinterpret_cast<QuantileState<double>*>(src.ptr);
+            return {static_cast<double>(quantile_state->get_value_by_percentile(percentile_value))};
+        } else {
+            QuantileState<double> quantile_state(Slice(src.ptr, src.len));
+            return {static_cast<double>(quantile_state.get_value_by_percentile(percentile_value))};
+        }
+
+
+    } else {
+        std::stringstream error_msg;
+        error_msg << "quantile_percent function's second argument must be constant. eg: quantile_percent(col, 0.95)";
+        ctx->set_error(error_msg.str().c_str()); 
+    }
+    return  DoubleVal::null();
+}
+
+StringVal QuantileStateFunctions::quantile_state_serialize(FunctionContext* ctx, const StringVal& src) {
+    if(src.is_null) {
+        return src;
+
+    }
+    auto tmp_ptr = reinterpret_cast<QuantileState<double>*>(src.ptr);
+    StringVal result = serialize(ctx, tmp_ptr);
+    delete tmp_ptr;
```

Comment:
Why delete src.ptr here? Where is its memory allocated?

Reply:
This function will be called in the PartitionedAggregationNode.
The memory is allocated from `quantile_state_init` initialize with a `empty quantile_state object`.
The `empty quantile_state` object is for aggregate calculations and it can be deleted after the aggregated results are serialized.